javascriptlib/src/mill/javascriptlib/PublishModule.scala

```diff
@@ -0,0 +1,539 @@
+package mill.javascriptlib
+
+import mill.*
+import os.*
+
+trait PublishModule extends TypeScriptModule {
+
+  /**
+   * Metadata about your project, required to publish.
+   *
+   * This is an equivalent of a `package.json`
+   */
+  def publishMeta: T[PublishModule.PublishMeta]
+
+  override def npmDevDeps: T[Seq[String]] =
+    Task { Seq("glob@^10.4.5", "ts-patch@3.3.0", "typescript-transform-paths@3.5.3") }
+
+  def bundledOut: T[String] = Task { "dist" }
+
+  private def declarationOut: T[String] = Task { "declarations" }
+
+  override def mainFileName: T[String] = Task { s"${millSourcePath.last}.js" }
+
+  // main file; defined with mainFileName
+  def main: T[String] =
+    Task { bundledOut() + "/src/" + mainFileName() }
+
+  private def mainType: T[String] = Task {
+    main().replaceFirst(bundledOut(), declarationOut()).replaceAll("\\.js", ".d.ts")
+  }
+
+  // Define exports for the package
+  // by default: mainFile is exported
+  // use this to define other exports
+  def exports: T[Map[String, String]] = Task { Map.empty[String, String] }
+
+  private def mkExports: T[Map[String, PublishModule.ExportEntry]] = Task {
+    exports().map { case (key, value) =>
+      key -> PublishModule.Export("./" + bundledOut() + "/" + value)
+    }
+  }
+
+  private def mkTypesVersion: T[Map[String, Seq[String]]] = Task {
+    pubAllSources().map { source =>
+      val dist = source.replaceFirst("typescript", bundledOut())
+      val declarations = source.replaceFirst("typescript", declarationOut())
+      ("./" + dist).replaceAll("\\.ts", "") -> Seq(declarations.replaceAll("\\.ts", ".d.ts"))
+    }.toMap
+  }
+
+  // build package.json from publishMeta
+  // mv to publishDir.dest
+  def packageJson: T[Unit] = Task { // PathRef
+    def splitDeps(input: String): (String, String) = {
+      input.split("@", 3).toList match {
+        case first :: second :: tail if input.startsWith("@") =>
+          ("@" + first + "@" + second, tail.mkString)
+        case first :: tail =>
+          (first, tail.mkString)
+      }
+    }
+
+    val json = publishMeta()
+    val updatedJson = json.copy(
+      files = json.files ++ Seq(bundledOut(), declarationOut()),
+      main = main(),
+      types = mainType(),
+      exports = Map("." -> PublishModule.Export("./" + main())) ++ mkExports(),
+      bin = json.bin.map { case (k, v) => (k, "./" + bundledOut() + "/" + v) },
+      typesVersions = mkTypesVersion(),
+      dependencies = transitiveNpmDeps().map { deps => splitDeps(deps) }.toMap,
+      devDependencies = transitiveNpmDeps().map { deps => splitDeps(deps) }.toMap
+    ).toJsonClean
+
+    os.write.over(publishDir().path / "package.json", updatedJson)
```

Comment:
You should never write to other task folders. You should instead return a `PathRef` to your own `Task.dest` folder and downstream tasks can copy and rename the files as necessary